quint/mempool/Consensus.qnt

```diff
@@ -0,0 +1,155 @@
+      val newProposed = state.proposed.set(p, txs)
+      state.with("proposed", newProposed)
+    }
+
+    // decide any proposed values
+    pure def decide(state: ConsensusState, p: Process): ConsensusState = {
+      val proposal = state.proposed.get(oneOf(state.proposed.keys().filter(x => state.proposed.get(x).size()>0)))
+      val decisions = state.decided.keys().filter(x => state.decided.get(x).size()>0)
+      val decision = if (decisions.size()>0) {state.decided.get(oneOf(decisions))}  else {proposal}
+      val newDecided = state.decided.set(p, decision)
+      state.with("decided",newDecided)
+    }
+
+    // INVARIANTS
+    
+    // Irrevocability: a process decides at most once
+    // This requires applying Consensus:mayDecide before Consensus:decide.
+    
+    // Validity: a process decides only submitted values
+    pure def consensusValidityInv(state: ConsensusState): bool = {
+      processes(state).forall(
+        p => not(hasDecided(state, p))
+	     or
+	     state.proposed.keys().filter(q => state.proposed.get(q) == state.decided.get(p)).size()>0
+      )
+    }
+    
+    // Agreement: no two processes decide different values
+    pure def consensusAgreementInv(state: ConsensusState): bool = {
+       processes(state).forall(x => processes(state).forall(y => setEqualsOrOneEmpty(state.decided.get(x), state.decided.get(y))))
+    }
+
+}
+
+module ConsensusTests {
+
+    import Spells.* from "Spells"
+    import System.* from "System"
+    import Consensus.*
+
+    pure val PROCESSES=Set("alice", "bob", "charlie")
+    pure val TXS=Set("T0", "T1", "T2", "T3")
+
+    var st: ConsensusState
+
+    val invariant: bool = {
+      consensusValidityInv(st) and consensusAgreementInv(st)
+    }
+
+    action init = all {
+        st' = newConsensus(PROCESSES)
+    }
+
+    action doPropose = all {
+      nondet p = oneOf(PROCESSES)
+      nondet txs = oneOf(TXS.powerset())
+      all {
+        require(mayPropose(st, p, txs)),
+        st' = propose(st, p, txs)
+      }
+    }
+
+    action doDecide = all {
+      nondet p = oneOf(PROCESSES)
+      all {
+        require(mayDecide(st, p)),
+    	st' = decide(st, p)
+      }
+    }
+
+    action step = any {
+      doPropose,
+      doDecide
+    }
+
+    run proposeTwiceError = {
+      nondet p = oneOf(PROCESSES)
+      nondet txs = oneOf(TXS.powerset())
+      init
+      .then(
+        all {
+    	  mayPropose(propose(st, p, txs), p, txs),
+    	  st' = st
+    	})
+      .fail()
+    }
+
+    run decideNonProposedError = {
+      init
+      .then(
+        all {
+  	  mayDecide(st, oneOf(PROCESSES)),
+	  st' = st
+	})
+      .fail()
+    }
+
+    run decideProposedSuccess = {
+      nondet p = oneOf(PROCESSES)
+      nondet q = oneOf(PROCESSES)
+      nondet txs0 = oneOf(TXS.powerset())
+      nondet txs1 = oneOf(TXS.powerset())
+      init
+      .then(st' = propose(propose(st, p, txs0), q, txs1))
```

Comment:
It is a matter of taste, I guess, but the following seems more readable.
```suggestion
      .then(st' = st.propose(p, txs0).propose(q, txs1))
```

Reply:
Indeed. I think python follows the same standard, in which the first parameter is the caller.

Reply:
I have changed everything this way to improve readability.